File: src-tauri/src/handlers/thumbnail.rs
use tauri::State;
use std::sync::Arc;
use std::path::{Path, PathBuf};
use crate::models::*;
use crate::db::{self};
use crate::core::ThumbnailGenerator;

/// 【Hash 直接拼接方案】获取缩略图，不存在则生成
/// 
/// 特点：
/// 1. 不查询数据库缩略图表，直接根据 hash 拼接路径检查文件
/// 2. 文件存在直接返回路径
/// 3. 文件不存在则调用生成服务创建缩略图
/// 
/// 特点：
/// 1. 不查询数据库缩略图表
/// 2. 直接根据 hash 拼接路径检查文件
/// 3. 文件不存在时调用生成服务创建缩略图
#[tauri::command]
pub async fn get_or_generate_thumbnail_by_hash(
    state: State<'_, crate::AppState>,
    image_id: i64,
    hash: String,
    image_path: String,
    size_type: String,
    thumbnail_dir: Option<String>,
) -> Result<Option<String>, String> {
    // 优先使用传入的自定义目录，否则使用当前设置目录
    let thumb_dir = if let Some(custom_dir) = thumbnail_dir {
        PathBuf::from(custom_dir)
    } else {
        state.thumbnail_dir.read().await.clone()
    };

    // 构造缩略图文件名：{hash}_{size}.jpg（缩略图统一使用 jpg 格式）
    let filename = format!("{}_{}.jpg", hash, size_type);
    let thumb_path = thumb_dir.join(&filename);

    // 如果文件已存在，直接返回路径（不再检查数据库，前端会缓存）
    if thumb_path.exists() {
        return Ok(Some(thumb_path.to_string_lossy().to_string()));
    }

    // 文件不存在，需要生成缩略图
    let size = ThumbnailSize::from_str(&size_type)
        .ok_or_else(|| format!("Invalid size type: {}", size_type))?;

    let result = state
        .thumbnail_service
        .generate_now(image_id, &image_path, &hash, size)
        .await
        .map_err(|e| e.to_string())?;

    if result.success {
        Ok(result.path)
    } else {
        tracing::warn!("Failed to generate thumbnail for hash {}: {:?}", hash, result.error);
        Ok(None)
    }
}

#[tauri::command]
pub async fn generate_thumbnail(
    state: State<'_, crate::AppState>,
    image_id: i64,
    size_type: String,
) -> Result<serde_json::Value, String> {
    // 获取图片信息
    let pool = state.db.lock().await.clone();
    let image = db::ImageRepository::get_by_id(&pool, image_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Image not found".to_string())?;

    // 解析尺寸类型
    let size = ThumbnailSize::from_str(&size_type)
        .ok_or_else(|| format!("Invalid size type: {}", size_type))?;

    // 生成缩略图
    let result = state
        .thumbnail_service
        .generate_now(image_id, &image.path, &image.hash, size)
        .await
        .map_err(|e| e.to_string())?;

    Ok(serde_json::json!({
        "success": result.success,
        "path": result.path,
        "width": result.width,
        "height": result.height,
        "error": result.error,
    }))
}

#[tauri::command]
pub async fn get_thumbnail_path(
    state: State<'_, crate::AppState>,
    image_id: i64,
    size_type: String,
) -> Result<Option<String>, String> {
    let pool = state.db.lock().await;

    // 先查询图片获取 hash
    let image = db::ImageRepository::get_by_id(&pool, image_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Image not found".to_string())?;

    let thumbnail = db::ThumbnailRepository::get_by_hash_and_size(&pool, &image.hash, &size_type)
        .await
        .map_err(|e| e.to_string())?;

    let thumb_dir = state.thumbnail_dir.read().await;
    Ok(thumbnail.map(|t| thumb_dir.join(&t.path).to_string_lossy().to_string()))
}

/// 【懒加载方案】获取缩略图路径，如果不存在则生成
/// 
/// 这个命令用于前端按需加载缩略图：
/// 1. 检查缩略图是否已存在
/// 2. 如果不存在，同步生成缩略图
/// 3. 返回缩略图路径（或 null 如果生成失败）
#[tauri::command]
pub async fn get_or_generate_thumbnail(
    state: State<'_, crate::AppState>,
    image_id: i64,
    size_type: String,
) -> Result<Option<String>, String> {
    let pool = state.db.lock().await.clone();
    let thumb_dir = state.thumbnail_dir.read().await.clone();

    // 获取图片信息
    let image = db::ImageRepository::get_by_id(&pool, image_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Image not found".to_string())?;

    // 1. 先检查缩略图是否已存在（使用 hash 查询）
    if let Some(thumbnail) = db::ThumbnailRepository::get_by_hash_and_size(&pool, &image.hash, &size_type)
        .await
        .map_err(|e| e.to_string())?
    {
        // 检查文件是否真实存在（DB 中存的是相对路径）
        let abs_path = thumb_dir.join(&thumbnail.path);
        if abs_path.exists() {
            return Ok(Some(abs_path.to_string_lossy().to_string()));
        }
        // 文件不存在，需要重新生成
    }

    // 2. 解析尺寸类型
    let size = ThumbnailSize::from_str(&size_type)
        .ok_or_else(|| format!("Invalid size type: {}", size_type))?;

    // 3. 生成缩略图
    let result = state
        .thumbnail_service
        .generate_now(image_id, &image.path, &image.hash, size)
        .await
        .map_err(|e| e.to_string())?;

    if result.success {
        Ok(result.path)
    } else {
        // 生成失败，返回 null（前端可以显示占位符）
        tracing::warn!("Failed to generate thumbnail for image {}: {:?}", image_id, result.error);
        Ok(None)
    }
}

#[tauri::command]
pub async fn get_thumbnail_status(
    state: State<'_, crate::AppState>,
    image_id: i64,
) -> Result<ThumbnailStatus, String> {
    let pool = state.db.lock().await;

    // 先查询图片获取 hash
    let image = db::ImageRepository::get_by_id(&pool, image_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Image not found".to_string())?;

    let mut status = db::ThumbnailRepository::get_thumbnail_status(&pool, &image.hash)
        .await
        .map_err(|e| e.to_string())?;

    // 将相对路径拼接为绝对路径
    let thumb_dir = state.thumbnail_dir.read().await;
    status.small_path = status.small_path.map(|p| thumb_dir.join(&p).to_string_lossy().to_string());
    status.medium_path = status.medium_path.map(|p| thumb_dir.join(&p).to_string_lossy().to_string());
    status.large_path = status.large_path.map(|p| thumb_dir.join(&p).to_string_lossy().to_string());

    Ok(status)
}

#[tauri::command]
pub async fn generate_all_thumbnails(
    state: State<'_, crate::AppState>,
    image_id: i64,
) -> Result<serde_json::Value, String> {
    let pool = state.db.lock().await.clone();

    // 获取图片信息
    let image = db::ImageRepository::get_by_id(&pool, image_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Image not found".to_string())?;

    // 获取当前缩略图目录
    let thumbnail_dir = state.thumbnail_dir.read().await.clone();
    let thumbnail_dir_arc = Arc::new(tokio::sync::RwLock::new(thumbnail_dir));

    let (generator, _progress_rx) = ThumbnailGenerator::new(pool.clone(), thumbnail_dir_arc, 2);
    let generator = Arc::new(generator);

    let results = generator
        .generate_all_sizes(image_id, &image.path, &image.hash)
        .await
        .map_err(|e| e.to_string())?;

    let success_count = results.iter().filter(|r| r.success).count();
    let fail_count = results.len() - success_count;

    Ok(serde_json::json!({
        "success": success_count == 3,
        "total": results.len(),
        "succeeded": success_count,
        "failed": fail_count,
        "results": results.iter().map(|r| {
            serde_json::json!({
                "size": r.task.size.as_str(),
                "success": r.success,
                "path": r.path,
                "width": r.width,
                "height": r.height,
            })
        }).collect::<Vec<_>>(),
    }))
}

#[tauri::command]
pub async fn check_thumbnails_integrity(
    state: State<'_, crate::AppState>,
) -> Result<serde_json::Value, String> {
    let pool = state.db.lock().await;
    let thumbnail_dir = state.thumbnail_dir.read().await.clone();

    // 获取所有缩略图记录（使用 image_hash）
    let records = sqlx::query_as::<_, (String, String, String, i64)>(
        "SELECT image_hash, size_type, path, file_size FROM thumbnails"
    )
    .fetch_all(&*pool)
    .await
    .map_err(|e| e.to_string())?;

    let record_count = records.len();
    let mut missing_files = Vec::new();
    let mut empty_files = Vec::new();
    let mut valid_count = 0;

    for (image_hash, size_type, path_str, _db_size) in records {
        // DB 中存的是相对路径，拼接为绝对路径再检查
        let abs_path = thumbnail_dir.join(&path_str);

        if !abs_path.exists() {
            missing_files.push(serde_json::json!({
                "image_hash": image_hash,
                "size_type": size_type,
                "path": abs_path.to_string_lossy().to_string()
            }));
        } else {
            match std::fs::metadata(&abs_path) {
                Ok(metadata) => {
                    let actual_size = metadata.len() as i64;
                    if actual_size == 0 {
                        empty_files.push(serde_json::json!({
                            "image_hash": image_hash,
                            "size_type": size_type,
                            "path": abs_path.to_string_lossy().to_string()
                        }));
                    } else {
                        valid_count += 1;
                    }
                }
                Err(e) => {
                    missing_files.push(serde_json::json!({
                        "image_hash": image_hash,
                        "size_type": size_type,
                        "path": format!("{} (read error: {})", abs_path.to_string_lossy(), e)
                    }));
                }
            }
        }
    }

    // 统计目录中的实际文件数
    let mut actual_file_count = 0;
    if let Ok(entries) = std::fs::read_dir(&thumbnail_dir) {
        for entry in entries.filter_map(|e| e.ok()) {
            let path = entry.path();
            if path.extension().map(|e| e == "jpg").unwrap_or(false) {
                actual_file_count += 1;
            }
        }
    }

    Ok(serde_json::json!({
        "summary": {
            "database_records": record_count,
            "actual_files": actual_file_count,
            "valid_files": valid_count,
            "missing_files": missing_files.len(),
            "empty_files": empty_files.len(),
        },
        "missing_files": missing_files,
        "empty_files": empty_files,
        "thumbnail_dir": thumbnail_dir.to_string_lossy().to_string(),
    }))
}

#[tauri::command]
pub async fn fix_missing_thumbnails(
    state: State<'_, crate::AppState>,
) -> Result<serde_json::Value, String> {
    let pool = state.db.lock().await.clone();
    let thumbnail_service = Arc::clone(&state.thumbnail_service);
    let thumbnail_dir = state.thumbnail_dir.read().await.clone();

    // 获取所有缩略图记录（使用 image_hash）
    let records = sqlx::query_as::<_, (String, String, String)> (
        "SELECT image_hash, size_type, path FROM thumbnails"
    )
    .fetch_all(&pool)
    .await
    .map_err(|e| e.to_string())?;

    let mut deleted_count = 0;
    let mut regenerated_count = 0;
    let mut failed_count = 0;
    let mut failed_list = Vec::new();

    for (image_hash, size_type, path_str) in records {
        // DB 中存的是相对路径，拼接为绝对路径
        let abs_path = thumbnail_dir.join(&path_str);

        // 检查文件是否存在且有效
        let needs_regen = if !abs_path.exists() {
            true
        } else {
            match std::fs::metadata(&abs_path) {
                Ok(metadata) => metadata.len() == 0,
                Err(_) => true,
            }
        };

        if needs_regen {
            // 删除无效的数据库记录（使用 hash）
            if let Err(e) = sqlx::query("DELETE FROM thumbnails WHERE image_hash = ?1 AND size_type = ?2")
                .bind(&image_hash)
                .bind(&size_type)
                .execute(&pool)
                .await
            {
                tracing::warn!("Failed to delete invalid thumbnail record: {}", e);
            }
            deleted_count += 1;

            // 获取原图信息（通过 hash 查找）
            match db::ImageRepository::get_by_hash(&pool, &image_hash).await {
                Ok(images) if !images.is_empty() => {
                    let image = &images[0];
                    let image_id = image.id;
                    
                    // 重新生成缩略图
                    let size = match size_type.as_str() {
                        "small" => ThumbnailSize::Small,
                        "medium" => ThumbnailSize::Medium,
                        "large" => ThumbnailSize::Large,
                        _ => ThumbnailSize::Small,
                    };

                    match thumbnail_service
                        .generate_now(image_id, &image.path, &image.hash, size)
                        .await
                    {
                        Ok(result) => {
                            if result.success {
                                regenerated_count += 1;
                                tracing::info!("Regenerated thumbnail for hash {} ({})", image_hash, size_type);
                            } else {
                                failed_count += 1;
                                failed_list.push(serde_json::json!({
                                    "image_hash": image_hash,
                                    "size_type": size_type,
                                    "error": result.error.unwrap_or_else(|| "Unknown error".to_string())
                                }));
                            }
                        }
                        Err(e) => {
                            failed_count += 1;
                            failed_list.push(serde_json::json!({
                                "image_hash": image_hash,
                                "size_type": size_type,
                                "error": e.to_string()
                            }));
                        }
                    }
                }
                Ok(_) => {
                    tracing::warn!("Image hash {} not found, cannot regenerate thumbnail", image_hash);
                    failed_count += 1;
                }
                Err(e) => {
                    tracing::warn!("Failed to get image by hash {}: {}", image_hash, e);
                    failed_count += 1;
                }
            }
        }
    }

    Ok(serde_json::json!({
        "deleted_records": deleted_count,
        "regenerated": regenerated_count,
        "failed": failed_count,
        "failed_list": failed_list,
    }))
}

File: src/api/thumbnail.ts
import { invoke, convertFileSrc } from "@tauri-apps/api/core";

const previewPathCache = new Map<string, string>();
const previewPromiseCache = new Map<string, Promise<string>>();

export type ThumbnailSize = "small" | "medium" | "large";

export interface ThumbnailStatus {
  image_id: number;
  has_small: boolean;
  has_medium: boolean;
  has_large: boolean;
  small_path: string | null;
  medium_path: string | null;
  large_path: string | null;
}

export interface ThumbnailResult {
  success: boolean;
  path: string | null;
  width: number | null;
  height: number | null;
  error: string | null;
}

export interface GenerateAllResult {
  success: boolean;
  total: number;
  succeeded: number;
  failed: number;
  results: {
    size: ThumbnailSize;
    success: boolean;
    path: string | null;
    width: number | null;
    height: number | null;
  }[];
}

/**
 * 生成单张图片的指定尺寸缩略图
 * @param imageId 图片ID
 * @param sizeType 尺寸类型: 'small' | 'medium' | 'large'
 */
export async function generateThumbnail(
  imageId: number,
  sizeType: ThumbnailSize,
  force?: boolean
): Promise<ThumbnailResult> {
  return await invoke<ThumbnailResult>("generate_thumbnail", {
    imageId,
    sizeType,
    force,
  });
}

/**
 * 获取指定图片和尺寸的缩略图路径
 * @param imageId 图片ID
 * @param sizeType 尺寸类型
 */
export async function getThumbnailPath(
  imageId: number,
  sizeType: ThumbnailSize
): Promise<string | null> {
  return await invoke<string | null>("get_thumbnail_path", {
    imageId,
    sizeType,
  });
}

/**
 * 【懒加载方案】获取缩略图路径，如果不存在则生成
 * 
 * 用于按需加载缩略图：
 * 1. 检查缩略图是否已存在
 * 2. 如果不存在，同步生成缩略图
 * 3. 返回缩略图路径（或 null 如果生成失败）
 * 
 * @param imageId 图片ID
 * @param sizeType 尺寸类型
 * @returns 缩略图路径，生成失败返回 null
 */
export async function getOrGenerateThumbnail(
  imageId: number,
  sizeType: ThumbnailSize
): Promise<string | null> {
  return await invoke<string | null>("get_or_generate_thumbnail", {
    imageId,
    sizeType,
  });
}

/**
 * 获取图片的缩略图状态
 * @param imageId 图片ID
 */
export async function getThumbnailStatus(
  imageId: number
): Promise<ThumbnailStatus> {
  return await invoke<ThumbnailStatus>("get_thumbnail_status", {
    imageId,
  });
}

/**
 * 为单张图片生成所有尺寸的缩略图
 * @param imageId 图片ID
 */
export async function generateAllThumbnails(
  imageId: number
): Promise<GenerateAllResult> {
  return await invoke<GenerateAllResult>("generate_all_thumbnails", {
    imageId,
  });
}

/**
 * 【Hash 直接拼接方案】获取缩略图，不存在则生成
 * 
 * @param imageId 图片ID
 * @param hash 图片哈希值
 * @param imagePath 原图路径
 * @param sizeType 尺寸类型
 * @param thumbnailDir 可选，自定义缩略图目录
 * @returns 缩略图路径，生成失败返回 null
 */
export async function getOrGenerateThumbnailByHash(
  imageId: number,
  hash: string,
  imagePath: string,
  sizeType: ThumbnailSize,
  thumbnailDir?: string
): Promise<string | null> {
  return await invoke<string | null>("get_or_generate_thumbnail_by_hash", {
    imageId,
    hash,
    imagePath,
    sizeType,
    thumbnailDir,
  });
}

/**
 * 获取缩略图URL（使用 Tauri 的 convertFileSrc）
 * @param path 缩略图路径
 */
export function getThumbnailUrl(path: string): string {
  // 使用 Tauri 的 convertFileSrc 将本地路径转换为可访问的 URL
  // 将 Windows 反斜杠转换为正斜杠
  return convertFileSrc(path.replace(/\\/g, '/'));
}

/**
 * 获取图片预览路径（HEIF/HEIC 会自动转换为 JPEG）
 * @param imagePath 原图路径
 * @returns 可用于前端显示的路径（HEIF 会返回转换后的 JPEG 路径）
 */
export async function getImagePreviewPath(imagePath: string): Promise<string> {
  // 非 HEIF 文件直接返回原路径
  const ext = imagePath.split('.').pop()?.toLowerCase();
  if (ext !== 'heif' && ext !== 'heic') {
    return imagePath;
  }
  
  // 检查缓存
  if (previewPathCache.has(imagePath)) {
    return previewPathCache.get(imagePath)!;
  }
  if (previewPromiseCache.has(imagePath)) {
    return previewPromiseCache.get(imagePath)!;
  }
  
  const promise = invoke<string>("get_image_preview_path", { imagePath });
  previewPromiseCache.set(imagePath, promise);
  
  try {
    const path = await promise;
    previewPathCache.set(imagePath, path);
    return path;
  } finally {
    previewPromiseCache.delete(imagePath);
  }
}

/**
 * 获取图片预览 URL（自动处理 HEIF/HEIC 转换）
 * @param imagePath 原图路径
 * @returns 可用于 img src 的 URL
 */
export async function getImagePreviewUrl(imagePath: string): Promise<string> {
  const path = await getImagePreviewPath(imagePath);
  return convertFileSrc(path.replace(/\\/g, '/'));
}

/**
 * 缩略图尺寸配置
 */
export const THUMBNAIL_SIZES: { [key in ThumbnailSize]: { label: string; size: number } } = {
  small: { label: "小图", size: 200 },
  medium: { label: "中图", size: 500 },
  large: { label: "大图", size: 1000 },
};
